apps/studio/src/server/modules/folder/folder.router.ts

```diff
@@ -130,33 +151,70 @@ export const folderRouter = router({
           userId: ctx.user.id,
         })
 
-        const result = await db
-          .updateTable("Resource")
-          .where("Resource.id", "=", resourceId)
-          .where("Resource.siteId", "=", Number(siteId))
-          .where("Resource.type", "in", [
-            ResourceType.Folder,
-            ResourceType.Collection,
-          ])
-          .set({
-            permalink,
-            title,
-          })
-          .returning(defaultFolderSelect)
-          .executeTakeFirst()
-          .catch((err) => {
-            if (get(err, "code") === PG_ERROR_CODES.uniqueViolation) {
-              throw new TRPCError({
-                code: "CONFLICT",
-                message: "A resource with the same permalink already exists",
-              })
-            }
-            throw err
+        const result = await db.transaction().execute(async (tx) => {
+          const user = await tx
+            .selectFrom("User")
+            .where("id", "=", ctx.user.id)
+            .selectAll()
+            .executeTakeFirstOrThrow(() => new TRPCError({ code: "NOT_FOUND" }))
+
+          const oldResource = await db
+            .selectFrom("Resource")
+            .selectAll()
+            .where("Resource.id", "=", resourceId)
+            .where("Resource.siteId", "=", Number(siteId))
+            .where("Resource.type", "in", [
+              ResourceType.Folder,
+              ResourceType.Collection,
+            ])
+            .executeTakeFirst()
+
+          if (!oldResource) {
+            throw new TRPCError({
+              code: "NOT_FOUND",
+              message: "Resource does not exist",
+            })
+          }
+
+          const newResource = await db
+            .updateTable("Resource")
+            .where("Resource.id", "=", resourceId)
+            .where("Resource.siteId", "=", Number(siteId))
+            .where("Resource.type", "in", [
+              ResourceType.Folder,
+              ResourceType.Collection,
+            ])
+            .set({
+              permalink,
+              title,
+            })
```

Comment:
since we already fetch the `oldResource`, should we find the newResource by using `oldResource.id`? otherwise the filtering logic is duplicated - in the future someone might update one without updating another one

Reply:
No opinions on this one, updated in 9fff24c507dd19fca1f26ccd2fba956c074169d8.